### ISSUE TITLE
Invalid header in slot_count on 32bit platforms only

### ISSUE THREAD

**Issue by erg** [factor/factor#2749]
There are a bunch of build emails that failed with `Invalid header in slot_count` but only on 32bit (linux, windows, mac).

<img width="797" alt="Screenshot 2023-01-24 at 10 06 25" src="https://user-images.githubusercontent.com/27430/214345284-a0464587-e675-45da-a735-cb41b018e945.png">

<img width="805" alt="Screenshot 2023-01-24 at 10 05 58" src="https://user-images.githubusercontent.com/27430/214345186-c773fae0-8dd9-4031-a81f-beb3107753ec.png">

<img width="798" alt="Screenshot 2023-01-24 at 10 06 52" src="https://user-images.githubusercontent.com/27430/214345388-13f18a51-d874-402d-9488-903b3948accd.png">

```
! linux
Loading resource:core/locals/macros/macros.factor
Loading resource:basis/bootstrap/layouts.factor
Building generic words...
Serializing words...
Serializing JIT data...
Serializing global namespace...
Serializing singletons...
Serializing special object table...
Performing word fixups...
Performing header fixups...
Image length: 563895
Object cache size: 70854
Writing image to /home/erg/builds32/2022-08-18-16-55/factor/boot.unix-arm.64.image...
You have triggered a bug in Factor. Please report.
critical_error: Invalid header in slot_count: 5007ff00
Starting low level debugger...
Basic commands:
  q ^D             -- quit Factor
  c                -- continue executing Factor - NOT SAFE
  t                -- throw exception in Factor - NOT SAFE
  .s .r .c         -- print data, retain, call stacks
  help             -- full help, including advanced commands

> ==== DATA STACK:
"benchmark.gc2"
==== RETAIN STACK:
#<bignum @ 0x8279a735>
{ "benchmark.3d-matrix-scalar" "benchmark.3d-matrix-vector" "benchmark.ant" "benchmark.backtrack" "benchmark.base64" "benchmark.beust1" "benchmark.beust2" "benchmark.binary-search" "benchmark.binary-trees" "benchmark.bloom-filters"... }


20 clients: 942787331
40 clients: 1901072005
100 clients: 5200939833
You have triggered a bug in Factor. Please report.
critical_error: Invalid header in slot_count: 5003ff00
Starting low level debugger...
Basic commands:
  q ^D             -- quit Factor
  c                -- continue executing Factor - NOT SAFE
  t                -- throw exception in Factor - NOT SAFE
  .s .r .c         -- print data, retain, call stacks
  help             -- full help, including advanced commands

> ==== DATA STACK:
"benchmark.yield"



Serializing special object table...
Performing word fixups...
Performing header fixups...
Image length: 563735
Object cache size: 70825
Writing image to /home/erg/builds32/2022-08-02-10-40/factor/boot.unix-arm.64.image...
You have triggered a bug in Factor. Please report.
critical_error: Invalid header in slot_count: 4fffff00
Starting low level debugger...
Basic commands:
  q ^D             -- quit Factor
  c                -- continue executing Factor - NOT SAFE
  t                -- throw exception in Factor - NOT SAFE
  .s .r .c         -- print data, retain, call stacks
  help             -- full help, including advanced commands

> ==== DATA STACK:
"benchmark.gc2"
==== RETAIN STACK:
#<bignum @ 0x8272a135>
{ "benchmark.3d-matrix-scalar" "benchmark.3d-matrix-vector" "benchmark.ant" "benchmark.backtrack" "benchmark.base64" "benchmark.beust1" "benchmark.beust2" "benchmark.binary-search" "benchmark.binary-trees" "benchmark.bloom-filters"... }


! windows

[master (root-commit) 883b24e] initial commit of empty file
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 empty-file
Factor 0.99 x86.32 (2168, heads/master-e1f7c71f63, Jun 21 2022 18:30:27)
[Microsoft Visual C++ 190023506] on windows
IN: scratchpad Factor 0.99 x86.32 (2168, heads/master-e1f7c71f63, Jun 21 2022 18:30:27)
[Microsoft Visual C++ 190023506] on windows
IN: scratchpad You have triggered a bug in Factor. Please report.
critical_error: Invalid header in slot_count: 7fffff04
Starting low level debugger...
Basic commands:
  q ^Z             -- quit Factor
  c                -- continue executing Factor - NOT SAFE
  t                -- throw exception in Factor - NOT SAFE
  .s .r .c         -- print data, retain, call stacks
  help             -- full help, including advanced commands

> ==== DATA STACK:
"benchmark.splay"




IN: scratchpad You have triggered a bug in Factor. Please report.
critical_error: Invalid header in slot_count: 7fffff04
Starting low level debugger...
Basic commands:
  q ^Z             -- quit Factor
  c                -- continue executing Factor - NOT SAFE
  t                -- throw exception in Factor - NOT SAFE
  .s .r .c         -- print data, retain, call stacks
  help             -- full help, including advanced commands

> ==== DATA STACK:
"benchmark.regex-dna"



! mac
Loading resource:core/math/floats/floats.factor
Loading resource:core/io/streams/c/c.factor
Loading resource:core/locals/fry/fry.factor
Loading resource:core/locals/macros/macros.factor
Loading vocab:bootstrap/layouts.factor
Building generic words...
Serializing words...
Serializing JIT data...
Serializing global namespace...
Serializing singletons...
Serializing special object table...
Performing word fixups...
Performing header fixups...
Image length: 565023
Object cache size: 71222
Writing image to /Users/erg/builds32/2021-03-12-13-10/factor/boot.unix-arm.64.image...
You have triggered a bug in Factor. Please report.
critical_error: Invalid header in slot_count: 3117ff00
Starting low level debugger...
Basic commands:
  q ^D             -- quit Factor
  c                -- continue executing Factor - NOT SAFE
  t                -- throw exception in Factor - NOT SAFE
  .s .r .c         -- print data, retain, call stacks
  help             -- full help, including advanced commands

> ==== DATA STACK:
"benchmark.knucleotide"
"GGTATT"




Serializing JIT data...
Serializing global namespace...
Serializing singletons...
Serializing special object table...
Performing word fixups...
Performing header fixups...
Image length: 565023
Object cache size: 71222
Writing image to /Users/erg/builds32/2021-03-12-19-40/factor/boot.unix-arm.64.image...
You have triggered a bug in Factor. Please report.
critical_error: Invalid header in slot_count: 310fff00
Starting low level debugger...
Basic commands:
  q ^D             -- quit Factor
  c                -- continue executing Factor - NOT SAFE
  t                -- throw exception in Factor - NOT SAFE
  .s .r .c         -- print data, retain, call stacks
  help             -- full help, including advanced commands

> ==== DATA STACK:
"benchmark.knucleotide"
```

**Comment by erg** [comment]
To fix this, we should find a way to crash it reliably. One idea is allocating some memory and doing a gc and running some of these benchmarks where it fails or `"kernel" reload` as in one of the failure emails.

The actual bug happens in `slot_visitor.hpp`:

```c++
////////// layouts.hpp
#define TAG_MASK 15
cell type() const { return (header >> 2) & TAG_MASK; }


// The number of slots (cells) in an object which should be scanned by
// the GC. The number can vary in arrays and tuples, in all other
// types the number is a constant.
template <typename Fixup>
inline cell object::slot_count(Fixup fixup) const {
  if (free_p())
    return 0;

  cell t = type();  // ************ corrupted
  if (t == ARRAY_TYPE) {
    // capacity + n slots
    return 1 + array_capacity((array*)this);
  } else if (t == TUPLE_TYPE) {
    tuple_layout* layout = (tuple_layout*)fixup.translate_data(
        untag<object>(((tuple*)this)->layout));
    // layout + n slots
    return 1 + tuple_capacity(layout);
  } else {
    switch (t) {
      // these objects do not refer to other objects at all
      case FLOAT_TYPE:
      case BYTE_ARRAY_TYPE:
      case BIGNUM_TYPE:
      case CALLSTACK_TYPE: return 0;
      case WORD_TYPE: return 8;
      case ALIEN_TYPE: return 2;
      case DLL_TYPE: return 1;
      case QUOTATION_TYPE: return 3;
      case STRING_TYPE: return 3;
      case WRAPPER_TYPE: return 1;
      default:
        critical_error("Invalid header in slot_count", (cell)this);
        return 0; // can't happen
    }
  }
}

enum type_tags {
  FIXNUM_TYPE, // 0
  F_TYPE,
  ARRAY_TYPE,
  FLOAT_TYPE,
  QUOTATION_TYPE,
  BIGNUM_TYPE,
  ALIEN_TYPE,
  TUPLE_TYPE,
  WRAPPER_TYPE,
  BYTE_ARRAY_TYPE,
  CALLSTACK_TYPE,
  STRING_TYPE,
  WORD_TYPE,
  DLL_TYPE, // 13

  TYPE_COUNT
};

```

We shift right by two and take the bottom four bits. So from an initial bit pattern of `abcdef` we shift off `ef` and `abcd` does not look up to one of the bit patterns 0..13, so it's either 14 or 15.

**Comment by erg** [comment]
I handle the `F_TYPE` now. If we see this bug, we know it's a 14 or a 15 and there is further corruption. I'm still not convinced it will be the fix, though, because why doesn't it break on 64bit?

**Comment by nomennescio** [comment]
That code hasn't change in at least 7 years (apart from your commit). Why does it crash suddenly? Doesn't it show that the object  is corrupted? An object pointer of `7fffff04` looks suspicious to me...

**Comment by mrjbq7** [comment]
Some of the slot visitor code changed a few years ago. I suspect a 64-bit assumption was made somewhere because it breaks on 32-bit, and only under heavy memory load. On Jan 25, 2023, at 10:15 AM, nomennescio ***@***.***> wrote:﻿
That code hasn't change in at least 7 years (apart from your commit). Why does it crash suddenly? Doesn't it show that the object  is corrupted? An object pointer of 7fffff04 looks suspicious to me...

**Comment by erg** [comment]
> That code hasn't changed in at least 7 years (apart from your commit). Why does it crash suddenly? Doesn't it show that the object is corrupted? An object pointer of `7fffff04` looks suspicious to me... Time for `git bisect`?

```
struct object {
  NO_TYPE_CHECK;
  // header format (bits indexed with least significant as zero):
  // bit 0      : free?
  // bit 1      : forwarding pointer?
  // if not forwarding:
  //   bit 2..5    : tag
  //   bit 6..end  : hashcode
  // if forwarding:
  //   bit 2..end  : forwarding pointer
  cell header;


0x7fffff04
bits 0, 1 are 0  -- not free, no forwarding ptr
2..5 => 0b0001 -- fixnum

0x7fffff04 -6 shift . ! hashcode
33554428

// maybe set by this?
void factor_vm::primitive_compute_identity_hashcode() {
  object* obj = untag<object>(ctx->pop());
  object_counter++;
  if (object_counter == 0)
    object_counter++;
  obj->set_hashcode((cell)obj ^ object_counter);
}
```

The best way to fix this would be to find a simple command to crash it reliably.

**Comment by mrjbq7** [comment]
Maybe this is fixed with the corruption issue ce31bc9ead1ee5c37e91339a9892ec316bbe9a85

**Comment by erg** [comment]
Did we audit all the codebase for these displaced aliens? I'm in favor of closing this til we see it again.
